File: apps/mobile/lib/shared/models/trip.dart
class TripMemberDto {
  const TripMemberDto({
    required this.userId,
    required this.role,
    required this.color,
    required this.ghostMode,
  });

  final String userId;
  final String role;
  final String color;
  final bool ghostMode;

  factory TripMemberDto.fromJson(Map<String, dynamic> json) => TripMemberDto(
        userId: json['user_id'] as String,
        role: json['role'] as String,
        color: json['color'] as String,
        ghostMode: json['ghost_mode'] as bool? ?? false,
      );
}

class TripDto {
  const TripDto({
    required this.id,
    required this.ownerId,
    required this.name,
    required this.status,
    required this.joinCode,
    required this.members,
    this.startAt,
    this.endAt,
  });

  final String id;
  final String ownerId;
  final String name;
  final String status;
  final String joinCode;
  final List<TripMemberDto> members;
  final DateTime? startAt;
  final DateTime? endAt;

  factory TripDto.fromJson(Map<String, dynamic> json) => TripDto(
        id: json['id'] as String,
        ownerId: json['owner_id'] as String,
        name: json['name'] as String,
        status: json['status'] as String,
        joinCode: json['join_code'] as String,
        startAt: json['start_at'] == null
            ? null
            : DateTime.parse(json['start_at'] as String),
        endAt: json['end_at'] == null
            ? null
            : DateTime.parse(json['end_at'] as String),
        members: ((json['members'] as List?) ?? const [])
            .map((m) => TripMemberDto.fromJson(m as Map<String, dynamic>))
            .toList(),
      );
}

File: apps/backend/app/schemas/trip.py
from __future__ import annotations

import uuid
from datetime import datetime, timezone

from pydantic import BaseModel, ConfigDict, Field, field_validator


class TripCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    start_at: datetime | None = None
    end_at: datetime | None = None

    @field_validator("start_at", "end_at")
    @classmethod
    def _as_utc(cls, value: datetime | None) -> datetime | None:
        # Assume UTC for naive input so start/end are always comparable — a
        # client mixing "…Z" and naive timestamps otherwise 500'd the
        # end - start subtraction ("can't subtract offset-naive and aware").
        if value is not None and value.tzinfo is None:
            return value.replace(tzinfo=timezone.utc)
        return value


class TripJoinRequest(BaseModel):
    join_code: str = Field(min_length=6, max_length=6)


class TripMemberOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: uuid.UUID
    role: str
    color: str
    ghost_mode: bool
    joined_at: datetime


class TripOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    owner_id: uuid.UUID
    name: str
    status: str
    start_at: datetime | None
    end_at: datetime | None
    join_code: str
    created_at: datetime
    members: list[TripMemberOut] = Field(default_factory=list)
